src/engine/interop/specification.md

```diff
@@ -0,0 +1,151 @@
+# Engine API. Interop edition
+
+This document specifies a subset of the Engine API methods that are required to be implemented for the Merge interop.
+
+*Note*: Sync API design is yet a draft and considered optional for the interop.
+
+## Underlying protocol
+
+This specification is based on [Ethereum JSON-RPC API](https://eth.wiki/json-rpc/API) and inherits the following properties of this standard:
+* Supported communication protocols (HTTP and WebSocket)
+* Message format and encoding notation
+* [Error codes improvement proposal](https://eth.wiki/json-rpc/json-rpc-error-codes-improvement-proposal)
+
+Client software **MUST** expose Engine API at a port independent from JSON-RPC API. The default port for the Engine API is 8550 for HTTP and 8551 for WebSocket.
+
+## Error codes
+
+The list of error codes introduced by this specification can be found below.
+
+| Code | Possible Return message | Description |
+| - | - | - |
+| 4 | Unknown block | Should be used when a call refers to the unknown block |
+| 5 | Unavailable payload | Should be used when the `payloadId` parameter of `engine_getPayload` call refers to the payload building process that is unavailable |
+
+## Structures
+
+Fields having `DATA` and `QUANTITY` types **MUST** be encoded according to the [HEX value encoding](https://eth.wiki/json-rpc/API#hex-value-encoding) section of Ethereum JSON-RPC API.
+
+*Note:* Byte order of encoded value having `QUANTITY` type is big-endian.
+
+### ExecutionPayload
+
+This structure maps on the [`ExecutionPayload`](https://github.com/ethereum/consensus-specs/blob/dev/specs/merge/beacon-chain.md#ExecutionPayload) structure of the beacon chain spec. The fields are encoded as follows:
+- `parentHash`: `DATA`, 32 Bytes
+- `coinbase`:  `DATA`, 20 Bytes
+- `stateRoot`: `DATA`, 32 Bytes
+- `receiptRoot`: `DATA`, 32 Bytes
+- `logsBloom`: `DATA`, 256 Bytes
+- `random`: `DATA`, 32 Bytes
+- `blockNumber`: `QUANTITY`, 64 Bits
+- `gasLimit`: `QUANTITY`, 64 Bits
+- `gasUsed`: `QUANTITY`, 64 Bits
+- `timestamp`: `QUANTITY`, 64 Bits
+- `extraData`: `DATA`, 0 to 32 Bytes
+- `baseFeePerGas`: `QUANTITY`, 256 Bits
+- `blockHash`: `DATA`, 32 Bytes
+- `transactions`: `Array of DATA` - Array of transaction objects, each object is a byte list (`DATA`) representing `TransactionType || TransactionPayload` or `LegacyTransaction` as defined in [EIP-2718](https://eips.ethereum.org/EIPS/eip-2718)
```

Comment:
Should this be specified as an RLP encoded list, i.e. `rlp([rlp(tx1), rlp(tx2), ...])`? If it isn't, I think it will take a non-conforming RLP decoder to read the list of txs.

Reply:
Isn't it just a json array of RLP bytes? The JSON RPC decoder iterates through and each item is passed to the RLP decoder for use on this layer?

1. it seems a bit unnatural for this type to be `transactions: DATA` where the DATA is an rlp list of rlp transactions. The consensus layer has a list (not RLP) of TXs so it's easiest to just pass that list in the transport's list encoding
2. this would require the consensus-layer to have an RLP encoder which would be very nice to avoid

Reply:
Not every item of this list is always an RLP-encoded string. It is RLP for legacy transactions and `TransactionType || TransactionPayload` for transactions which are encoded according to EIP-2718.

Reply:
My mistake, I realize now this is a _JSON_ array of serialized txs.